### PR TITLE
Enable hybrid parallelisation

### PR DESCRIPTION
**Pull request by aradi**
* Makefiles modified to enable hybrid (MPI/OpenMP) parallelisation by default.
* Keyword UseOmpThreads added, so that users must request hybrid parallelisation explicitely.
* Socket handling fixed to avoid long timeouts blocking the port.

### CODE REVIEW

**Comment by aradi** [comment]
Note: buildbot tests are still running, so do not merge yet. But it is open for discussions. :wink:

File: INSTALL.rst [hunk 1]
```diff
@@ -97,21 +97,30 @@ Compile
   build directory location within the `make.config` file (variable
   ``BUILDDIR``), but the default location is inside the root of the source tree.
 
+
+Test
```
Comment (bhourahine) [inline INSTALL.rst]: Testing DFTB+
============

File: INSTALL.rst [hunk 2]
```diff
@@ -97,21 +97,30 @@ Compile
...
 * After successful compilation, execute the tests with ::
 
     make test
 
   You may also run the tests in parallel (option ``-j``) in order to speed this
   up.  If you use parallel testing, ensure that the number of OpenMP threads is
-  set to be 1 via the ``OMP_NUM_THREADS`` environment variable before starting
-  the tests, e.g.::
+  reduced accordingly. As an example, assuming 4 cores in your workstation you
+  could issue::
```
Comment (bhourahine) [inline INSTALL.rst]: could use::

File: doc/dftb+/manual/dftbp.tex [hunk 3]
```diff
@@ -3012,6 +3013,20 @@ \section{Parallel}
   of MPI-processes. Default: 1 (all processes work at the same k-point and spin
   at a given time).
 
+\item[\is{UseOmpThreads}] Enables the usage of OpenMP-threads (hybrid
+  MPI/OpenMP-parallelisation). In order to prevent you from accidently running
+  more processes and threads than appropriate for your hardware, this feature is
+  turned off by default. Consequently the MPI-parallelised binary stops if the
```
Comment (bhourahine) [inline doc/dftb+/manual/dftbp.tex]: Consequently in this case an MPI-parallelised binary will stop

File: doc/dftb+/manual/dftbp.tex [hunk 4]
```diff
@@ -3012,6 +3013,20 @@ \section{Parallel}
...
+  maximal number of OpenMP-threads is greater than one when DFTB+ is
+  started. (You can usually set the number of maximally allowed OpenMP-threads
+  by setting the \is{OMP\_NUM\_THREADS} environment variable in your shell.)
+
+  You can enable this option if you wish to run DFTB+ with hybrid
+  parallelisation. You would then typically start less MPI-processes than
```
Comment (bhourahine) [inline doc/dftb+/manual/dftbp.tex]: fewer MPI-processes

File: doc/dftb+/manual/dftbp.tex [hunk 5]
```diff
@@ -3012,6 +3013,20 @@ \section{Parallel}
...
+  physical cores on each node and set the number of threads accordingly.  This
```
Comment (bhourahine) [inline doc/dftb+/manual/dftbp.tex]: and also set the number

**Comment by aradi** [comment]
I updated now to use updated mpifx with `mpi_init_thread`. Currently from my personal repo, but once the PR for mpifx is merged, I'll change it back to the official one. Let me know, if you see any problems.

**Pull request merged**